proposed/psr-ai.md

```diff
@@ -0,0 +1,252 @@
+6. The PSR MUST NOT prescribe any particular provider API, HTTP format, or runtime behavior.
+
+## 5. The AI Client Abstraction
+
+### 5.1. The Client Interface
+
+The ClientInterface represents the entry point for AI interaction.
+
+An implementation MUST provide at least the following methods:
+
+```php
+namespace Psr\Ai;
+
+interface ClientInterface
+{
+    /**
+     * @param MessageInterface|array<MessageInterface> $message
+     */
+    public function complete(MessageInterface|array $message): ResponseInterface;
+
+    /**
+     * @param MessageInterface|array<MessageInterface> $message
+     */
+    public function stream(MessageInterface|array $message): StreamInterface;
+
+    public function invokeTool(ToolInterface $tool, array $arguments): mixed;
+}
+```
+
+**Rules:**
+- `complete()` MUST perform a synchronous model call and return a finalized `ResponseInterface`.
+- `stream()` MUST return a `StreamInterface` that yields incremental output as available.
+- `invokeTool()` MUST execute a callable tool in the same process or over MCP.
+
+### 5.2. The Message Interface
+
+The MessageInterface defines a structured input or output exchanged between client and model.
+
+```php
+namespace Psr\Ai;
+
+interface MessageInterface
+{
+    public function role(): string;
+    public function content(): string|array;
+    public function metadata(): array;
+}
+```
+
+**Rules**:
+- `role()` MUST return a string such as system, user, or assistant.
+- `content()` MAY contain plain text or structured data (e.g., JSON, array).
+- Implementations MAY define message classes (e.g., `UserMessage`, `AssistantMessage`) for typed interactions.
+- Message instances MUST be immutable once created.
+
+### 5.3. The Stream Interface
+
+The StreamInterface enables incremental consumption of model output.
+
+```php
+namespace Psr\Ai;
+
+interface StreamInterface extends \Traversable
+{
+    public function onToken(callable $callback): void;
+    public function onError(callable $callback): void;
+    public function onComplete(callable $callback): void;
+}
+```
+
+**Rules:**
+- Implementations MUST emit partial output (e.g., tokens, chunks) through `onToken()`.
+- `onError()` MUST emit any transport or model errors.
+- `onComplete()` MUST be called when the stream has ended.
+- Streams SHOULD be compatible with PSR-7 stream semantics where applicable.
+
+### 5.4. The Tool Interface 
+
+The ToolInterface standardizes callable functions the model may request via tool calling.
+
+```php
+namespace Psr\Ai;
+
+interface ToolInterface
+{
+    public function name(): string;
+    public function description(): ?string;
+    public function schema(): array;
+    public function execute(array $arguments): mixed;
+}
+```
+
+**Rules:**
+- Tools MUST define their input/output schema for validation and serialization.
+- Implementations SHOULD align schema definitions with the [Model Context Protocol](https://modelcontextprotocol.io/).
```

Comment:
What do you mean by this? In general (with a few exceptions) APIs and MCP expects jsonapi. We should codify that the schema is an array that represents valid jsonapi and allow serializers to handle any idiosyncrasies.